crates/mun_language_server/src/main_loop.rs

```diff
@@ -1,26 +1,220 @@
-use crate::protocol::{Connection, Message};
+use crate::analysis::{Analysis, AnalysisSnapshot};
+use crate::change::AnalysisChange;
+use crate::config::{Config, FilesWatcher};
+use crate::conversion::{convert_range, url_from_path_with_drive_lowercasing};
+use crate::protocol::{Connection, Message, Notification, Request, RequestId};
 use crate::Result;
-use futures::StreamExt;
+use anyhow::anyhow;
+use async_std::sync::RwLock;
+use futures::channel::mpsc::{unbounded, Sender, UnboundedReceiver, UnboundedSender};
+use futures::{SinkExt, StreamExt};
+use lsp_types::notification::PublishDiagnostics;
+use lsp_types::{PublishDiagnosticsParams, Url};
+use ra_vfs::{RootEntry, Vfs, VfsChange, VfsFile};
+use serde::{de::DeserializeOwned, Serialize};
+use std::collections::HashSet;
+use std::sync::Arc;
 
+/// A `Task` is something that is send from async tasks to the entry point for processing. This
+/// enables synchronizing resources like the connection with the client.
+#[derive(Debug)]
+enum Task {
+    Notify(Notification),
+}
+
+#[derive(Debug)]
 enum Event {
     Msg(Message),
+    Vfs(ra_vfs::VfsTask),
+    Task(Task),
+}
+
+/// State for the language server
+struct LanguageServerState {
```

Comment:
It seems like the `LanguageServerState` consists of the current `LanguageServerSnapshot` plus the `vfs_task_receiver`. Would it be an idea to use composition instead of duplication?

Reply:
@tdejager ?

Reply:
It doesnt contain exactly the same info. `Analysis` vs `AnalysisSnapshot`